src/v/datalake/protobuf_to_arrow_converter.cc

```diff
@@ -0,0 +1,190 @@
+    }
+
+    std::vector<std::shared_ptr<arrow::Field>> field_vec;
+    for (auto& [field_idx, array] : _arrays) {
+        auto field = message_desc->field(field_idx);
+        field_vec.push_back(array->field(field->name()));
+    }
+    return field_vec;
+}
+std::shared_ptr<arrow::Schema>
+datalake::proto_to_arrow_converter::build_schema() {
+    return arrow::schema(build_field_vec());
+}
+void datalake::proto_to_arrow_converter::add_message_parsed(
+  std::unique_ptr<google::protobuf::Message> message) {
+    // TODO: Allocating and deallocating the field descriptor array
+    // for every message is probably a bad idea.
+    auto reflection = message->GetReflection();
+    std::vector<const google::protobuf::FieldDescriptor*> field_descriptors;
+    reflection->ListFields(*message, &field_descriptors);
+    for (auto& [field_idx, array] : _arrays) {
+        assert(
+          field_idx >= 0); // Dummy assert to avoid unused variable warning.
+        // TODO: handle this error
+        if (!array->add_value(message.get(), field_idx).ok()) {
+            // FIXME: exception type
+            throw new std::runtime_error("Could not add value");
+        }
+    }
+}
+void datalake::proto_to_arrow_converter::initialize_protobuf_schema(
+  const std::string& schema) {
+    google::protobuf::io::ArrayInputStream proto_input_stream(
+      schema.c_str(), schema.size());
+    google::protobuf::io::Tokenizer tokenizer(&proto_input_stream, nullptr);
+
+    google::protobuf::compiler::Parser parser;
+    if (!parser.Parse(&tokenizer, &_file_descriptor_proto)) {
+        // TODO: custom exception type, or does something exist in wasm or
+        // schema registry already?
+        throw std::runtime_error("Could not parse protobuf schema");
+    }
+
+    if (!_file_descriptor_proto.has_name()) {
+        _file_descriptor_proto.set_name("dummy_message_name");
+    }
+
+    _file_desc = _protobuf_descriptor_pool.BuildFile(_file_descriptor_proto);
+    if (_file_desc == nullptr) {
+        throw std::runtime_error("Could not build descriptor pool");
+    }
+}
+bool datalake::proto_to_arrow_converter::initialize_arrow_arrays() {
+    using namespace detail;
+    namespace pb = google::protobuf;
+
+    const pb::Descriptor* message_desc = message_descriptor();
+    if (message_desc == nullptr) {
+        std::cerr << "Message descriptor is null\n";
+        return false;
+    }
+    // TODO: This code is almost identical to the code in proto_to_arrow_struct.
+    // It seems like we could combine these. One possibility is to tread every
+    // message as a struct, and parse everything with proto_to_arrow_struct, so
+    // this class is only responsible for parsing the schema and setting that
+    // up.
+    for (int field_idx = 0; field_idx < message_desc->field_count();
+         field_idx++) {
+        auto desc = message_desc->field(field_idx);
+
+        if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_INT32) {
```

Comment:
seems like the same loop above w/ the same code, one just pushes back the other move-assigns but the return type could be a free function w/ tests.